src/RadixDlt.NetworkGateway.PostgresIntegration/Models/EntityResourceVaultAggregateHistory.cs

```diff
@@ -84,6 +89,107 @@ internal class EntityResourceVaultAggregateHistory
+    /// This property is nullable in order to ensure smooth database migration process.
+    /// </remarks>
+    [Column("vault_significant_update_state_versions")]
+    public List<long>? VaultSignificantUpdateStateVersions { get; set; }
+
+    public static EntityResourceVaultAggregateHistory Create(long databaseId, long entityId, long resourceEntityId, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, null, entityId, resourceEntityId, stateVersion);
+    }
+
+    public static EntityResourceVaultAggregateHistory CopyOf(long databaseId, EntityResourceVaultAggregateHistory other, long stateVersion)
+    {
+        return CopyOrCreate(databaseId, other, other.EntityId, other.ResourceEntityId, stateVersion);
+    }
+
+    /// <summary>
+    /// Attempts to add new or update existing vault unless it already is the most recently modified one.
+    /// </summary>
+    /// <returns>true if added or modified, otherwise false.</returns>
+    public bool TryUpsert(long vaultId, long stateVersion)
+    {
+        EnsureVaultSignificantUpdateStateVersionsExists();
+
+        var currentIndex = VaultEntityIds.IndexOf(vaultId);
+
+        // we're already the most recent one and there's no point potential second resource with the same state version
+        if (currentIndex == 0 && VaultEntityIds.Count == 1)
+        {
+            return false;
+        }
+
+        // we're already the most recent one but there might be second resource with matching state_version
+        // so in order to guarantee that we'll be unambiguously earlier in the overall sequence we might potentially
+        // want to update anyways
+        if (currentIndex == 0 && VaultEntityIds.Count > 1 && VaultSignificantUpdateStateVersions[1] < VaultSignificantUpdateStateVersions[0])
+        {
+            return false;
+        }
+
+        if (currentIndex != -1)
+        {
+            VaultEntityIds.RemoveAt(currentIndex);
+            VaultSignificantUpdateStateVersions.RemoveAt(currentIndex);
+        }
+
+        VaultEntityIds.Insert(0, vaultId);
+        VaultSignificantUpdateStateVersions.Insert(0, stateVersion);
+
+        return true;
+    }
+
+    public bool ShouldBePersisted()
+    {
+        if (_originalVaultEntityIds == null)
+        {
+            return true;
+        }
+
+        if (!_originalVaultEntityIds.SequenceEqual(VaultEntityIds.ToArray()))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    private static EntityResourceVaultAggregateHistory CopyOrCreate(long databaseId, EntityResourceVaultAggregateHistory? other, long entityId, long resourceEntityId, long stateVersion)
+    {
+        var ret = new EntityResourceVaultAggregateHistory
+        {
+            Id = databaseId,
+            FromStateVersion = stateVersion,
+            EntityId = entityId,
+            ResourceEntityId = resourceEntityId,
+            VaultEntityIds = new List<long>(other?.VaultEntityIds.ToArray() ?? Array.Empty<long>()),
```

Comment:
This feels a little indirect? Could we not do `other?.VaultEntityIds.ToList() ?? List.Empty<long>()`?

Reply:
There's no `List.Empty<long>` in C#. We could optionally move copy part to the if statement (`ret.VaultEntityIds.AddRange(other.ValueEntitiesIds)`)

Reply:
Ahh sorry, `other?.VaultEntityIds.ToList() ?? new List<long>()` - but this isn't a biggy :D